File: Code/Java/BinaryTreeZigzagLevelOrderTraversal.java
// Solution to https://leetcode.com/problems/binary-tree-zigzag-level-order-traversal/

class Solution {
    public List<List<Integer>> zigzagLevelOrder(TreeNode root) {
        List<List<Integer>> ans = new ArrayList<>();
        if (root == null) {
            return ans;
        }
        Queue<TreeNode> queue = new LinkedList<>();
        boolean leftToRight = true;
        queue.add(root);
        while (!queue.isEmpty()) {
            int size = queue.size();
            List<Integer> temp = new ArrayList<>();
            while (size-- > 0) {
                TreeNode removed = queue.remove();
                temp.add(removed.val);
                if (removed.left != null) {
                    queue.add(removed.left);
                }
                if (removed.right != null) {
                    queue.add(removed.right);
                }
            }
            if (!leftToRight) {
                Collections.reverse(temp);
            }
            ans.add(temp);
            leftToRight = !leftToRight;
        }
        return ans;
    }
}

File: DataStructures/Binary_tree_zigzag_level_order_traversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<vector<int>> zigzagLevelOrder(TreeNode* root) {
       vector<vector<int>> v;
        if(root==NULL)
            return v;
        
	// Code here
	stack<TreeNode *> s1;
	stack<TreeNode *> s2;
	s1.push(root);
	
        vector<int> a;
	while(!s1.empty()||!s2.empty())
	{  while(!s1.empty()){
	    TreeNode *temp=s1.top();
	    s1.pop();
        
	    a.push_back(temp->val);
	   
	   if(temp->left)s2.push(temp->left); 
	   if(temp->right)s2.push(temp->right);
	      
	}
     
      if(a.size()!=0) 
       {  v.push_back(a);
     a.clear();
       }  
	    
	while(!s2.empty()){
	    TreeNode *temp=s2.top();
	    s2.pop();
	    a.push_back(temp->val);
	    
	    
	        
	  if(temp->right)s1.push(temp->right);
	   if(temp->left)s1.push(temp->left);    
	        
	}
	   if(a.size()!=0) 
       {  v.push_back(a);
     a.clear();
       }  
	    
	}
	   return v;	// Code here

        
           
        
        
    }
};
